### DISCUSSION TITLE
Does This Work on Rollup Chains (Optimism, Arbitrum)?

### DISCUSSION THREAD

**Discussion by PaulRBerg** [pcaversaccio/xdeployer#2] [Q&A] [answered]
Pardon me for opening an issue for a simple question (may be worth it to enable [discussions](https://docs.github.com/en/discussions) on this repo?)

I'm currently using the [deterministic-deployment-proxy](https://github.com/Zoltu/deterministic-deployment-proxy) to deploy contracts deterministic addresses across chains. One of the issues with that [it can't be used on Arbitrum and Optimism](https://github.com/Zoltu/deterministic-deployment-proxy/issues/7)..

Is xdeployer able to deploy to rollup chains as well?

**Comment by pcaversaccio** [comment, marked as answer]
@paulrberg good point: I enabled discussions [here](https://github.com/pcaversaccio/xdeployer/discussions). 

Regarding your question: yes `xdeployer` is able to deploy on rollup chains. Currently, however, only the testnets are supported (I will add later the prod networks):
- [`optimismtestnet`](https://kovan-optimistic.etherscan.io)
- [`arbitrumtestnet`](https://testnet.arbiscan.io)

For testing purposes, I just deployed a mock [`ERC20`](https://github.com/pcaversaccio/xdeployer/blob/main/test/fixture-projects/rinkeby-kovan-project-with-constructor/contracts/ERC20Mock.sol) including constructors on both rollup chains:
- `optimismtestnet`: [`MyToken`](https://kovan-optimistic.etherscan.io/token/0xbdeeb8208465c507e489dfc98702a08f7c2a2d03)
- `arbitrumtestnet`: [`MyToken`](https://testnet.arbiscan.io/token/0xbdeeb8208465c507e489dfc98702a08f7c2a2d03)

![image](https://user-images.githubusercontent.com/25297591/141683037-8836fe24-1f55-4f32-93e7-6cd3388ea22a.png)

Also (just for testing) I also transferred the tokens and it worked smoothly:
![image](https://user-images.githubusercontent.com/25297591/141683194-c1bf5cb4-8113-4d23-8913-3ef717911200.png)
![image](https://user-images.githubusercontent.com/25297591/141683195-efcaad11-b605-484b-a722-2819329b7f2f.png)

However, you must increase the `gasLimit` significantly to deploy on Arbitrum as it seems. Also, in case interested check [here](https://github.com/OffchainLabs/arbitrum/blob/master/packages/arb-bridge-peripherals/contracts/tokenbridge/arbitrum/StandardArbERC20.sol) for the standard Arbitrum ERC-20 interface.

**Comment by PaulRBerg** [comment]
> I enabled discussions here.

Note that you can convert this issue to a discussion. There's a button on the right-hand side of the screen.

> However, you must increase the `gasLimit` significantly to deploy on Arbitrum as it seems

Yes, we have to do that because of the way the rollup chain txs are batched. That was actually my fear, that bumping the gas limit would break xdeployer.

**Reply by pcaversaccio** [reply]
> Note that you can convert this issue to a discussion. There's a button on the right-hand side of the screen.

Thanks, done 😀.

> Yes, we have to do that because of the way the rollup chain txs are batched. That was actually my fear, that bumping the gas limit would break deployer.

No, it does not break `xdeployer`. Currently, there is an upper acceptable `gasLimit` of 15m (due to the Ethereum mainnet) that the plugin checks for. If this threshold is exceeded, the plugin will not start any deployment. You can set your own `gasLimit` via the `xdeployer` configuration. Let me know if you think 15m is not enough for your application.

**Comment by 0xD006F00D** [comment]
Hi @pcaversacci, the proxy Paul linked is actually just to deploy the create2 contract itself and he stated the gasLimit is influencing the deployment address. So to ask the question differently - How did you deploy the create2 contract itself?

**Reply by pcaversaccio** [reply]
I have a section in the [`README`](https://github.com/pcaversaccio/xdeployer#how-it-works) on how this works. In order to save the deployment of the `CREATE2` helper contract itself, I deployed the helper smart contract on all EVM chains supported. Here is the repository: [create2deployer](https://github.com/pcaversaccio/create2deployer). This will save deployment costs for anyone that uses this plugin as well as keeps the complexity low. Please see [here](https://github.com/pcaversaccio/xdeployer/discussions/34) for certain design decisions regarding governance modifiers.

**Reply by 0xD006F00D** [reply]
So you just used the [deploy script](https://github.com/pcaversaccio/create2deployer/blob/main/scripts/deploy.ts) to deploy the create2deployer contract to every chain? Was the deployment nonce always the same when you deployed it or how come the create2deployer address is the same across networks?

**Reply by pcaversaccio** [reply]
Yes exactly using my deployer address `0x554282Cf65b42fc8fddc6041eb24ce5E8A0632Ad` with the same nonce on all chains. It's similar to the approach taken at [`multicall`](https://github.com/mds1/multicall).
